Irv ballots as ordered list

vteague opened the pull request:
Updates IRV storage so that IRV votes are an ordered list of plan names ("Alice", "Bob", "Chuan", "Diego") rather than a list with explicit parentheses ("Alice (1)", "Chuan (3)" "Bob (2)", "Diego (4)").

Checks whether the input CVR row is a valid IRV preference list (without skipped or repeated preferences or repeated candidates), and throws an error back to the user if it isn't.

Still not perfect in a variety of ways:
- it doesn't acually delete the explicit-parentheses choices from the list of available choices. I'm not sure whether this is a bug or a feature.
- it doesn't seem to parse our example file with write-ins properly. Not certain whether this is because the parser is wrong or the file is invalid.

vteague committed:
Started sketching IRV ballot interpretation at the CSV parse point.

vteague committed:
Merge pull request #54 from DemocracyDevelopers/scratch

Scratch

vteague committed:
Added some code and tests for IRV ballot interpretation as a list of (candidate name) strings.

vteague committed:
Update choice-parsing code so that it thinks the choices are "Alice", "Bob" not "Alice(1)" "Alice(2)" "Bob(1)" "Bob(2)" etc.

vteague committed:
Updating contest choices seems to be working. CVRs not yet updated.

vteague committed:
Working on updating CVRs with new votes (without explicit ranks in parentheses).

vteague committed:
IRV parsing now working for complicated candidate names.

vteague committed:
Removed IRVCountyContestResult.java

vteague committed:
Cleaner IRV contest rewriting

vteague committed:
Cleaner IRV contest rewriting

vteague committed:
Cleaned up default CountyContestResult values for IRV.

vteague committed:
IRV CSV parsing working. Still need to check details for write-ins. Error messages are not necessarily particularly helpful.

vteague committed:
Reorganized some IRV classes in model.

vteague committed:
Comments about some better ways to do it.

vteague committed:
Altered it so that the _contest_ choices remain as "Alice (1)", "Alice (2)" etc, but the _CVR_ choices are an ordered list of plain names.

michelleblom commented on server/eclipse-project/src/main/java/us/freeandfair/corla/csv/DominionCVRExportParser.java:
```diff
@@ -37,9 +39,14 @@
 import us.freeandfair.corla.util.ExponentialBackoffHelper;
+import us.freeandfair.corla.util.IRVParsingException;
+
+import static us.freeandfair.corla.model.ContestType.IRV;
```
Rather than importing the specific enumeration case ContestType.IRV, just import ContestType and then use ContestType.IRV where necessary.

michelleblom commented on server/eclipse-project/src/main/java/us/freeandfair/corla/csv/DominionCVRExportParser.java:
```diff
@@ -369,7 +376,7 @@ private Result addContests(final CSVRecord choiceLine,
         // "Write-in" is a fictitious candidate that denotes the beginning of
         // the list of qualified write-in candidates
-        final boolean isFictitious = "Write-in".equalsIgnoreCase(choice);
+        final boolean isFictitious = "Write-in".equalsIgnoreCase(choice) || IsIRVWriteIn(choice);
```
What is the purpose of IsIRVWriteIn()?

michelleblom commented on server/eclipse-project/src/main/java/us/freeandfair/corla/model/CVRContestInfo.java:
```diff
@@ -110,14 +110,16 @@ public CVRContestInfo(final Contest the_contest, final String the_comment,
     my_consensus = the_consensus;
     my_choices.addAll(the_choices);
+    boolean isIRV = my_contest.description().equalsIgnoreCase(ContestType.IRV.toString());
+
```
What happens when an additional type of contest is considered? With this model there are If IRV then .. in many classes.

michelleblom reviewed (changes requested):
Let's look at alternate design options that don't require changes to so many classes, and to keep parsing functionality bundled within the dominion parsing class.

vteague committed:
Much cleaner IRV parsing without much impact on classes other than DominionCVRExportParser.java.

This now stores both Contest Choices and CVR choices as plain names (without parentheses).

vteague committed:
Deleting unused code.

vteague committed:
Deleting unused code.

vteague committed:
Deleting unused code.

vteague committed:
Deleting unused code.

vteague committed:
Checking for correct exception in tests.
